File: packages/javascript/lib/array/arrIntersection.ts
import { comparableKey } from './_comparableKey.js';

export function arrIntersection(...arrays: any[][]): any[] {
	if (arrays.length < 1 || !Array.isArray(arrays[0])) {
		return [];
	}

	const [first, ...rest] = arrays;
	const otherKeys: Array<Set<string>> = [];

	for (let i = 0, restLength = rest.length; i < restLength; i += 1) {
		// A non-array argument shares no value with anything, so the result is empty.
		if (!Array.isArray(rest[i])) {
			return [];
		}

		otherKeys.push(new Set(rest[i].map(comparableKey)));
	}

	const seen = new Set<string>();
	const result: any[] = [];

	for (let i = 0, firstLength = first.length; i < firstLength; i += 1) {
		const value = first[i];
		const key = comparableKey(value);

		if (seen.has(key)) {
			continue;
		}

		seen.add(key);

		if (otherKeys.every((keys) => keys.has(key))) {
			result.push(value);
		}
	}

	return result;
}

File: packages/python/qsu/array/arrIntersection.py
from ._comparableKey import _comparableKey


def arrIntersection(*arrays) -> list:
	if len(arrays) < 1 or not isinstance(arrays[0], (list, tuple)):
		return []

	otherKeys = []

	for other in arrays[1:]:
		# A non-array argument shares no value with anything, so the result is empty.
		if not isinstance(other, (list, tuple)):
			return []

		otherKeys.append({_comparableKey(value) for value in other})

	seen = set()
	result = []

	for value in arrays[0]:
		key = _comparableKey(value)

		if key in seen:
			continue

		seen.add(key)

		if all(key in keys for keys in otherKeys):
			result.append(value)

	return result
